### PR TITLE
TKC TKL_A/B87: Initial support

### PR DESCRIPTION
**Pull request by TerryMathews**
## Description

Initial support for TKC TKL A/B87 PCB. Availability and boards TBA.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by drashna — APPROVED** [review]
Could you take a look at the PR Checklist and apply the suggestions from there?

**Comment by TerryMathews** [comment]
@drashna See if that covers what you were looking for.

File: keyboards/tkc/tkl_ab87/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+}
+
+void matrix_scan_user(void) {
+
+}
+
+void led_set_user(uint8_t usb_led) {
+
+}
```
Comment (zvecr) [inline keyboards/tkc/tkl_ab87/keymaps/default/keymap.c:44]:
```suggestion
```

File: keyboards/tkc/tkl_ab87/keymaps/via/keymap.c [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+}
+
+void matrix_scan_user(void) {
+
+}
+
+void led_set_user(uint8_t usb_led) {
+
+}
```
Comment (zvecr) [inline keyboards/tkc/tkl_ab87/keymaps/via/keymap.c:80]:
```suggestion
```

File: keyboards/tkc/tkl_ab87/readme.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# TKC TKL A/B87
+
+![Photo](https://i.imgur.com/BRHfEbH.jpg)
+
+A fairly typical TKL
+
+Keyboard Maintainer: TerryMathews  
+Hardware Supported: TKC TKL A/B87 PCB
+Hardware Availability: TBA
```
Comment (zvecr) [inline keyboards/tkc/tkl_ab87/readme.md:9]:
```suggestion
* Keyboard Maintainer: TerryMathews
* Hardware Supported: TKC TKL A/B87 PCB
* Hardware Availability: TBA
```

File: keyboards/tkc/tkl_ab87/rules.mk [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# MCU name
+MCU = atmega32u4
+
+BOOTLOADER = atmel-dfu
```
Comment (zvecr) [inline keyboards/tkc/tkl_ab87/rules.mk:4]:
```suggestion
# Bootloader selection
BOOTLOADER = atmel-dfu
```

File: keyboards/tkc/tkl_ab87/tkl_ab87.c [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+#include "tkl_ab87.h"
+
+bool led_update_kb(led_t led_state) {
+    bool res = led_update_user(led_state);
+    if(res) {
+        writePin(F1, led_state.caps_lock);
+        writePin(F0, led_state.scroll_lock);
+    }
+    return res;
+}
```
Comment (zvecr) [inline keyboards/tkc/tkl_ab87/tkl_ab87.c:25]:
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/tkc/tkl_ab87/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [0] = LAYOUT_all( /* Base */
+    
+	KC_ESC,           KC_F1,   KC_F2,   KC_F3,   KC_F4,        KC_F5,   KC_F6,   KC_F7,   KC_F8,       KC_F9,   KC_F10,  KC_F11,  KC_F12,           KC_PSCR, KC_SLCK, KC_PAUS, \
+	KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_DEL, KC_BSPC,           KC_INS,  KC_HOME, KC_PGUP, \
+	KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC,          KC_BSLS,          KC_DEL,  KC_END,  KC_PGDN, \
+	KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_NUHS,          KC_ENT,                                      \
+	KC_LSFT, KC_NUBS, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT, KC_DEL,                    KC_UP,            \
+	KC_LCTL, KC_LGUI, KC_LALT,                           KC_SPC,                                       KC_RALT, KC_RGUI, KC_APP,  KC_RCTL,          KC_LEFT, KC_DOWN, KC_RGHT  \
```
Comment (fauxpark) [inline keyboards/tkc/tkl_ab87/keymaps/default/keymap.c:26]:
```suggestion
	KC_ESC,           KC_F1,   KC_F2,   KC_F3,   KC_F4,        KC_F5,   KC_F6,   KC_F7,   KC_F8,       KC_F9,   KC_F10,  KC_F11,  KC_F12,           KC_PSCR, KC_SLCK, KC_PAUS,
	KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_DEL, KC_BSPC,           KC_INS,  KC_HOME, KC_PGUP,
	KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC,          KC_BSLS,          KC_DEL,  KC_END,  KC_PGDN,
	KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_NUHS,          KC_ENT,
	KC_LSFT, KC_NUBS, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT, KC_DEL,                    KC_UP,
	KC_LCTL, KC_LGUI, KC_LALT,                           KC_SPC,                                       KC_RALT, KC_RGUI, KC_APP,  KC_RCTL,          KC_LEFT, KC_DOWN, KC_RGHT
```

File: keyboards/tkc/tkl_ab87/keymaps/via/keymap.c [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    
+	_______,           _______,   _______,   _______,   _______,        _______,   _______,   _______,   _______,       _______,   _______,  _______,  _______,           _______, _______, _______, \
+	_______,  _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,  _______, _______,           _______,  _______, _______, \
+	_______,  _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,          _______,          _______,  _______,  _______, \
+	_______, _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______, _______,          _______,                                      \
+	_______, _______, _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,  _______,          _______, _______,                    _______,            \
+	_______, _______, _______,                           _______,                                       _______, _______, _______,  _______,          _______, _______, _______  \
```
Comment (fauxpark) [inline keyboards/tkc/tkl_ab87/keymaps/via/keymap.c:62]:
```suggestion
	KC_ESC,           KC_F1,   KC_F2,   KC_F3,   KC_F4,        KC_F5,   KC_F6,   KC_F7,   KC_F8,       KC_F9,   KC_F10,  KC_F11,  KC_F12,           KC_PSCR, KC_SLCK, KC_PAUS,
	KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_DEL, KC_BSPC,           KC_INS,  KC_HOME, KC_PGUP,
	KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC,          KC_BSLS,          KC_DEL,  KC_END,  KC_PGDN,
	KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_NUHS,          KC_ENT,
	KC_LSFT, KC_NUBS, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT, KC_DEL,                    KC_UP,
	KC_LCTL, KC_LGUI, KC_LALT,                           KC_SPC,                                       KC_RALT, KC_RGUI, KC_APP,  KC_RCTL,          KC_LEFT, KC_DOWN, KC_RGHT
  ),
  [FUNCTION] = LAYOUT_all( /* Base */
    
	_______,           _______,   _______,   _______,   _______,        _______,   _______,   _______,   _______,       _______,   _______,  _______,  _______,           _______, _______, _______,
	_______,  _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,  _______, _______,           _______,  _______, _______,
	_______,  _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,          _______,          _______,  _______,  _______,
	_______, _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______, _______,          _______,
	_______, _______, _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,  _______,          _______, _______,                    _______,
	_______, _______, _______,                           _______,                                       _______, _______, _______,  _______,          _______, _______, _______
  ),
  [ALTERNATE] = LAYOUT_all( /* Base */
    
	_______,           _______,   _______,   _______,   _______,        _______,   _______,   _______,   _______,       _______,   _______,  _______,  _______,           _______, _______, _______,
	_______,  _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,  _______, _______,           _______,  _______, _______,
	_______,  _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,          _______,          _______,  _______,  _______,
	_______, _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______, _______,          _______,
	_______, _______, _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,  _______,          _______, _______,                    _______,
	_______, _______, _______,                           _______,                                       _______, _______, _______,  _______,          _______, _______, _______ 
  ),
  [LAST] = LAYOUT_all( /* Base */
    
	_______,           _______,   _______,   _______,   _______,        _______,   _______,   _______,   _______,       _______,   _______,  _______,  _______,           _______, _______, _______,
	_______,  _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,  _______, _______,           _______,  _______, _______,
	_______,  _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,          _______,          _______,  _______,  _______,
	_______, _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______, _______,          _______,
	_______, _______, _______,    _______,    _______,    _______,    _______,    _______,    _______,    _______, _______,  _______,          _______, _______,                    _______,
	_______, _______, _______,                           _______,                                       _______, _______, _______,  _______,          _______, _______, _______
```

File: keyboards/tkc/tkl_ab87/readme.md [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+A fairly typical TKL
+
+* Keyboard Maintainer: TerryMathews
+* Hardware Supported: TKC TKL A/B87 PCB
+* Hardware Availability: TBA
+
+Make example for this keyboard (after setting up your build environment):
+
+    make tkc/tkl_ab87:default:flash
```
Comment (fauxpark) [inline keyboards/tkc/tkl_ab87/readme.md:13]:
```suggestion
    make tkc/tkl_ab87:default

Flashing example for this keyboard:

    make tkc/tkl_ab87:default:flash
```

File: keyboards/tkc/tkl_ab87/rules.mk [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output on port C6
```
Comment (fauxpark) [inline keyboards/tkc/tkl_ab87/rules.mk:23]:
```suggestion
BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

File: keyboards/tkc/tkl_ab87/keymaps/via/keymap.c [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+#include QMK_KEYBOARD_H
+
+//Layers
+
+enum {
+  BASE = 0,
+  FUNCTION,
+  ALTERNATE,
+  LAST,
+};
```
Comment (Erovia) [inline keyboards/tkc/tkl_ab87/keymaps/via/keymap.c:25]:
```suggestion
enum {
  BASE,
  FUNCTION,
};
```
4 layers for a TKL with 32u4 is **really** pushing it.
I'd suggest going with 2 or 3 layers.

**Review by Erovia — CHANGES_REQUESTED** [review]

File: keyboards/tkc/tkl_ab87/readme.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+* Hardware Availability: TBA
+
+Make example for this keyboard (after setting up your build environment):
+
+    make tkc/tkl_ab87:default
+
+Flashing example for this keyboard:
+
+    make tkc/tkl_ab87:default:flash
+
```
[the remark this answers is not in the document]
Comment (TerryMathews) [reply keyboards/tkc/tkl_ab87/readme.md:25]: Added.

File: keyboards/tkc/tkl_ab87/keymaps/via/keymap.c (hunk 10 above)
Comment (TerryMathews) [reply keyboards/tkc/tkl_ab87/keymaps/via/keymap.c]: Reducing to two VIA keymaps took a few more changes than just this, but I pushed a commit that took care of it. FWIW, I tested the firmware on real hardware and it didn't crash.

File: keyboards/tkc/tkl_ab87/readme.md [hunk 12]
```diff
@@ -11,6 +11,13 @@ A fairly typical TKL
 Make example for this keyboard (after setting up your build environment):
 
     make tkc/tkl_ab87:default
+	
+Reset board in one of two ways:
+
+	1) Short reset header to right of up arrow key
+	![Photo](https://i.imgur.com/S01CTTG.jpg)
+	
+	2) Hold down Esc key while plugging in USB-C
```
Comment (Erovia) [inline keyboards/tkc/tkl_ab87/readme.md]: I'd rather have the make and flashing example in one block and the reset instructions below or above it.

Comment (TerryMathews) [reply keyboards/tkc/tkl_ab87/readme.md]: Done.

File: keyboards/tkc/tkl_ab87/readme.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
+# TKC TKL A/B87
+
+![Photo](https://i.imgur.com/BRHfEbH.jpg)
```
Comment (noroadsleft) [inline keyboards/tkc/tkl_ab87/readme.md:3]: Swaps the image for a smaller version.

```suggestion
![Photo](https://i.imgur.com/BRHfEbHl.jpg)
```

File: keyboards/tkc/tkl_ab87/readme.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+A fairly typical TKL
+
+* Keyboard Maintainer: TerryMathews
```
Comment (noroadsleft) [inline keyboards/tkc/tkl_ab87/readme.md:7]: Turns this into a link.

```suggestion
* Keyboard Maintainer: [TerryMathews](https://github.com/TerryMathews)
```

File: keyboards/tkc/tkl_ab87/readme.md [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+* Hardware Supported: TKC TKL A/B87 PCB
+* Hardware Availability: TBA
+
+	
+Reset board in one of two ways:
+
+	1) Short reset header to right of up arrow key
+![Photo](https://i.imgur.com/S01CTTG.jpg)
```
Comment (noroadsleft) [inline keyboards/tkc/tkl_ab87/readme.md:15]: Smaller version.

```suggestion
![Photo](https://i.imgur.com/S01CTTGl.jpg)
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Review by noroadsleft — APPROVED** [review]

**Review by Erovia — APPROVED** [review]

**Pull request merged**